Add publication status for articles

Ydrasil opened the pull request:
Now we have to think about how to implement this feature for user experience
Cc @vinyll

Ydrasil committed:
Add publication status for articles

Ydrasil committed:
Add publication status for articles

vinyll commented on blog/test.py:
```diff
@@ -38,3 +49,18 @@ def test_language_detection(self):
+        self.assertIn(b'<article class=preview>', result.data)
+
+    def test_unpublished_article_does_not_appear_in_articles(self):
+        article = Article(title='<h1>title<br></h1>', content='<p>content</p>')
```
That's a bug you are raising here. We should not have `<br>` in the title

vinyll commented on blog/models.py:
```diff
@@ -39,6 +39,7 @@ class Article(db.Document):
     author = db.ReferenceField(User)
     language = db.StringField(min_length=2, max_length=2, default='en')
     images = db.ListField()
+    is_published = db.BooleanField()
```
I guess we could use a `publication_date` instead of a `is_published`.

`is_published` could then be a property:

```python
@property
def is_published():
    return bool(self.publication_date)
```

Ydrasil replied:
I have a question about this.
I'll explain this with an example.
If I publish an article, then I unpublish this article, then I publish the article again, what will be the date of publication ? The first time I've published it or the last time ?

vinyll replied:
@Ydrasil I would say it's the first day of publication.
We have the `modification_date` for other purposes.

vinyll commented on user/test.py:
```diff
@@ -0,0 +1,46 @@
+        User.objects.delete()
+        Article.objects.delete()
+
+    def test_unpublished_articles_does_not_appears_if_the_author_not_logged(
```
I believe you could move any test about article in the _blog_ module.
Let's imagine the project would someday see the _blog_ module removed,  these tests should be removed as well. (not planned hu, just exposing modules responsibilities).

vinyll commented on user/test.py:
```diff
@@ -0,0 +1,46 @@
+        article.author = self.user
+        article.is_published = False
+        article.save()
+        result = self.client.get('/@none')
```
Should we actually have a user with _@none_ as slug?

vinyll reviewed (changes requested):
That's a well written feature with tests, good job again 👍 
I believe it would be better if you were working with @fuentesloic for implementing the frontend as well.

It wouldn't make sense to implement _dead code_ (code that is not used); we try to remove any code that is not used! ;)

vinyll commented:
@fuentesloic: Shall we make an appointment to work together with @Ydrasil?

vinyll commented on user/test.py:
```diff
@@ -1,19 +1,33 @@
         self.client = app.test_client()
         self.user = User(email='emailtest@test.com').set_password('testtest')
+        self.user.allow_localization = True
         self.user.save()
```
```python
self.user = (User(email='emailtest@test.com',
                allow_location=True)
                .set_password('testtest')
                ).save()
```

vinyll commented on user/test.py:
```diff
@@ -44,3 +58,28 @@ def test_unpublished_articles_does_appears_if_the_author_is_logged(self):
+        result = self.client.get('/privacy')
+        self.assertEqual(result.status_code, 401)
+
+    def test_privacy_page(self):
```
a more explicit name like `def test_privacy_page_access(self)`?

vinyll reviewed (commented):
Will be good for merge after the small code improvements and the frontend implementation :)

Ydrasil committed:
Merge branch 'master' into publication-status

fuentesloic commented:
I'm on it!
Sorry for the late! :)

fuentesloic committed:
Around - form to add poi (#108)

fuentesloic committed:
init publish button

fuentesloic committed:
init publish button

fuentesloic committed:
if edit condition in article to publish button

fuentesloic committed:
if edit condition in article to publish button

fuentesloic committed:
remove quote

fuentesloic committed:
remove quote

fuentesloic committed:
copy script to change behaviour button

fuentesloic committed:
copy script to change behaviour button

fuentesloic committed:
move checkbox inside form create

fuentesloic committed:
move checkbox inside form create

fuentesloic committed:
add publish field in edit

fuentesloic committed:
add publish field in edit

fuentesloic committed:
fix erase error

fuentesloic committed:
fix erase error

fuentesloic committed:
put publish button in article profile

fuentesloic committed:
put publish button in article profile

vinyll commented on templates/blog/macros.html:
```diff
@@ -17,10 +17,12 @@
       {% endif %}
+      {% if edit %}
       <form>
         <input name="published" id="published" type="checkbox">
```
no quotes required here

vinyll commented on templates/blog/macros.html:
```diff
@@ -17,10 +17,12 @@
+      {% if edit %}
       <form>
         <input name="published" id="published" type="checkbox">
         <label for="published" title="draft or publish">published</label>
```
neither here

vinyll commented on templates/blog/macros.html:
```diff
@@ -19,8 +19,8 @@
       <form>
-        <input name="published" id="published" type="checkbox">
-        <label for="published" title="draft or publish">published</label>
+        <input name=published type=checkbox>
```
you removed the `id` but the label's `for` was refering to it

vinyll commented on templates/user/profile.html:
```diff
@@ -55,8 +55,8 @@ <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
-      <input name="published" id="published" type="checkbox">
-      <label for="published" title="draft or publish">published</label>
+      <input name=published type=checkbox>
+      <label for='published' title=draft or publish>published</label>
```
no need of quotes, even single quotes.
The `for` links to an `id`. You just deleted the id…

Ydrasil committed:
Add is_published to article

vinyll commented on templates/user/profile.html:
```diff
@@ -73,6 +73,12 @@ <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
+  
+  const publishField = document.querySelector('[name=published]')
+  publishField.addEventListener('change', event => {
+    apiFeedback(call, 'Your article is publish preference has been updated.')
```
It could be nice to have a message that mentions the state:
"You article has been published." or "You article has been unpublished."

vinyll replied:
We could display a message that would be contextual:

'You article has been published.'
'You article has been unpublished.'

fuentesloic committed:
publishField method

fuentesloic committed:
publishField method

vinyll commented on templates/user/profile.html:
```diff
@@ -73,6 +73,12 @@ <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
+  publishField.addEventListener('change', event => {
+    apiFeedback(call, 'Your article is publish preference has been updated.')
+  })
+  publishField.addEventListener('change', event => event.target.checked)
```
What is this supposed to do?

vinyll replied:
What is this supposed to do?

vinyll commented on templates/user/profile.html:
```diff
@@ -54,10 +54,6 @@ <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
     <h1><span class="newtext title">&nbsp;</span></h1>
     <p><span class=newtext>&nbsp;</span><span class=newtext>&nbsp;</span><span class=newtext>&nbsp;</span></p>
```
WTF is this?

vinyll commented on user/views.py:
```diff
@@ -14,10 +14,14 @@
+        if user == current_user:
+            articles = Article.objects(author=user)
+        else:
+            articles = Article.objects(author=user, is_published__ne=False)
```
```python
articles = Article.objects(author=user)
if user == current_user:
    articles = articles.filter(is_published__ne=False)
```

Should do the trick in a shorter way ;)

fuentesloic committed:
fix url post into form publish

fuentesloic committed:
fix url post into form publish

vinyll commented on user/test.py:
```diff
@@ -52,3 +56,21 @@ def test_delete_trip(self):
+
+    def test_unpublished_articles_does_appears_if_the_author_is_logged(self):
+        login_user(self)
+        Article(title='<h1>Article that must appear<br></h1>',
```
no html please, it must be cleaned out anyways (and you have a dedicated test for that) ;)

vinyll commented on user/test.py:
```diff
@@ -52,3 +56,21 @@ def test_delete_trip(self):
+        Article(title='<h1>Article that must appear<br></h1>',
+                content='<p>content</p>', author=self.user,
+                is_published=False).save()
+        result = self.client.get('/@emailtest')
```
ending slash

vinyll reviewed (changes requested):
You may also want to test to access an article with it's url if it's **un**published.

vinyll commented:
We will still probably need the publication date.
It may be a good option to replace `is_published` with `publication_date` and have it to `None` when it's unpublished.

vinyll commented on templates/blog/macros.html:
```diff
@@ -33,6 +33,7 @@
       </figure>
       {{ article.content|striptags|truncate(200) }}
+    {% set url = url_for_article(article) %}
     <a href={{ url }} lang={{ article.language }} class='read action-button'>Read more</a>
```
Using `{% set url = url_for_article(article) %}` is great when you need to reuse `url` multiple times.

Here you could just say: `<a href={{ url_for_article(article) }}`

fuentesloic replied:
oh true ! my bad

vinyll commented on templates/user/profile.html:
```diff
@@ -82,7 +82,7 @@ <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
   const publishField = document.querySelector('[name=publish]')
   publishField.addEventListener('change', event => {
-    const call = api.send('post', "{{ url_for('profile_edit') }}",
+    const call = api.send('post', event.target.form.action,
```
👍

fuentesloic replied:
;) I learned with someone pretty good 💯 / 💯

fuentesloic commented on templates/blog/article.html:
```diff
@@ -60,11 +60,15 @@ <h1 name=title>{{ article.title or 'Insert your title here' }}</h1>
     <input type=hidden name=title>
     <input type=hidden name=content>
+    <div>
+      <input name=publish type=checkbox>
```
need an id

fuentesloic commented on templates/blog/macros.html:
```diff
@@ -17,13 +17,23 @@
+      {% set url = url_for('article_edit', id=article.id) %}
+      <form method=post action={{ url }} edit>
+        <div>
+          <input name="publish" id="publish" type="checkbox">
```
need a specific id (with the article.id maybe)

fuentesloic commented on templates/user/profile.html:
```diff
@@ -79,6 +80,15 @@ <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
   })
   localizeField.addEventListener('change', event => event.target.checked && localize())
 
+  const publishField = document.querySelector('[name=publish]')
```
use querySelector with id

vinyll replied:
why a querySelecot with id?

fuentesloic replied:
because if I use by `name` that just take the firstbecause if I use by `name` that just take the first

vinyll replied:
Do you have multiple time the `name=publish` in your document?

fuentesloic replied:
yes on each article in you profile

fuentesloic replied:
but don't succeed to fix it ... in each case the ID block me because I want a string without number or sign

vinyll replied:
You're not publishing from your profile but from the article page, right?

vinyll replied:
mhhh… I see we're publishing an article from templates/user/profile.html.
Why not publishing an article from the article's page?

fuentesloic replied:
when you create an article it's right
But after that, I was wondering that user could be publish/unpublish article from is profile view

vinyll replied:
Well the issue you're on is supposed to solve [editing article's meta data](https://github.com/neomad-team/neomad.org/issues/27).
Changing the status of an article would be interested to be done in a list if a user needed to have a massive status change for multiple articles, but the purpose here is to be able to change non visible informations of an article (like the publication status, publication date, and later other infos such as the slug, transfer the owner and so on).
Hence the interest of having a status bar.

I'm running this version on my local copy to check the points of views ;)

vinyll replied:
I've completed user tests with the following feedback: https://github.com/neomad-team/neomad.org/pull/115#issuecomment-324452994

fuentesloic committed:
fix url for read more button

Commit:
Merge branch 'master' into publication-status

vinyll commented:
As the owner of the article https://dev.neomad.org/@vinyll/developpeur-sans-domicile-fixe-58faa70b1675bb01b46fcfa4/, when editing my article, I have a toolbar in the bottom for saving or deleting my article.
It could be a good place where to offer extra options, like (un)publishing, or changing the publication date.

![screen shot 2017-08-23 at 10 22 12 pm](https://user-images.githubusercontent.com/145172/29636682-d3b20b60-8851-11e7-87b8-3d1fafddc6e9.png)

We could have either the easy version (replacing the _save_ button with 2 buttons: _save as draft_ and _save and publish_), or an icon featuring a more advanced sub-menu with options (like we have for the user menu, see the screenshot below)


![screen shot 2017-08-23 at 10 27 14 pm](https://user-images.githubusercontent.com/145172/29636798-4d2e058e-8852-11e7-858f-54be56df33c5.png)
)

fuentesloic commented:
@vinyll it's already the case

fuentesloic committed:
add evolutive ID in input

fuentesloic committed:
Merge branch 'publication-status' of https://github.com/Ydrasil/neomad.org into publication-status

vinyll commented on templates/blog/article.html:
```diff
@@ -66,7 +66,8 @@ <h1 name=title>{{ article.title or 'Insert your title here' }}</h1>
     <div>
-      <input name=publish type=checkbox>
+      {% set id=article.id %}
+      <input name=publish type=checkbox class=publish id=publish{{ id }}>
```
is this id thing still useful?

vinyll commented on templates/blog/macros.html:
```diff
@@ -19,7 +19,8 @@
         <div>
-          <input name="publish" id="publish" type="checkbox">
+          {% set id=article.id %}
+          <input name=publish type=checkbox class=publish id="publish{{ id }}">
```
same question

fuentesloic committed:
transmit data into form

vinyll commented on blog/test.py:
```diff
@@ -165,3 +165,18 @@ def test_language_detection(self):
+        article.is_published = False
+        article.author = self.user
+        article.save()
+        result = self.client.get('/articles')
```
ending slash

vinyll commented on blog/test.py:
```diff
@@ -165,3 +165,18 @@ def test_language_detection(self):
+        article = Article(title='title', content='<p>content</p>')
+        article.author = self.user
+        article.save()
+        result = self.client.get('/articles')
```
ending slash

vinyll commented on templates/blog/article.html:
```diff
@@ -60,12 +60,14 @@ <h1 name=title>{{ article.title or 'Insert your title here' }}</h1>
     <input type=hidden name=content>
-    <button type=submit class=save>Save</button>
+    <input type=hidden name=is_published>
+    <button type=submit class=save value="">Save as draft</button>
```
change the color of draft for a gray background?

vinyll commented on templates/blog/macros.html:
```diff
@@ -22,7 +22,7 @@
       </figure>
       {{ article.content|striptags|truncate(200) }}
-    <a href={{ url }} lang={{ article.language }} class='read action-button'>Read more</a>
+    <a href={{ url_for_article(article) }} lang={{ article.language }} class='read action-button'>Read more</a>
```
double quotes for html when quotes are required

fuentesloic replied:
so href requires quotes I think it's the same for `src` in `img` (isn't?)

vinyll replied:
yes, the rule is simple:

1. don't use quotes unless required
2. when required, use double quotes

fuentesloic replied:
how you know if it's require ? like I know for class I need it when we have a space inside ... same for src.
There are other attribute ?

vinyll replied:
You know the rule, same for any attribute: if there's a caracter that would break the html in the value of the attribute, you need quotes. Such as a space, a `=`, a `<`, ` >, ` `/`, a `'` …

vinyll commented on user/test.py:
```diff
@@ -52,4 +56,23 @@ def test_delete_trip(self):
+
+    def test_unpublished_articles_does_not_appears_if_the_author_not_logged(
+            self):
+        Article(title='<h1>Article that must not appear<br></h1>',
```
remove html in title

vinyll commented on user/test.py:
```diff
@@ -52,4 +56,23 @@ def test_delete_trip(self):
+        Article(title='<h1>Article that must not appear<br></h1>',
+                content='<p>content</p>', author=self.user,
+                is_published=False).save()
+        result = self.client.get('/@emailtest')
```
ending slash

vinyll commented:
This PR seems almost good.
The only thing would be to store this `publication_date` instead of the `is_published` as a boolean.

When the article is published, we should save the current (`datetime.datetime.utcnow()`) date as publication_date.
When the article is saved as draft, the publication_date should be `None`.

Ydrasil committed:
Rebase publication-status from master

fuentesloic committed:
handle quote in html

fuentesloic committed:
draft ribbon responsive in profile article list

fuentesloic committed:
draft ribbon responsive in edit article

fuentesloic committed:
remove useless footer when editing article

fuentesloic committed:
remove last commit

fuentesloic committed:
fix side effect in preview in home page

fuentesloic committed:
add is_published condition into core.py

fuentesloic commented:
I push some design stuff here because that was quick (and I add condition into home previews article):
- When you edit:
![capture d ecran 2017-08-25 a 10 31 16-fullpage](https://user-images.githubusercontent.com/8755623/29719033-a8a36c52-8982-11e7-81a0-5dd156ce3512.png)
- In your article list:
![capture d ecran 2017-08-25 a 10 47 24-fullpage](https://user-images.githubusercontent.com/8755623/29719109-dd462a44-8982-11e7-90e0-99abb1af5c1f.png)

vinyll commented:
The draft ribbon seems appropriate to me. Just wondering about design details: should we go for flat design and remove the linear gradient? Should we make the ribbon width thinner?

fuentesloic commented:
@vinyll I should change design for be closer to the actual design (with color). Let me do an other try.

vinyll commented:
Right. That will probably get into the styleguide that @newick is working. Which – by the way – is going to shake some ass!!! 🚀

fuentesloic committed:
flatty design

fuentesloic commented:
New flatty design purpose, font lighter, using neomad link color
![capture d ecran 2017-08-25 a 13 50 41-fullpage](https://user-images.githubusercontent.com/8755623/29726260-0b22fad6-899d-11e7-84e0-aed9a6dd39b3.png)

Ydrasil committed:
WIP: Add publication_date

- POST '/article/<id>/edit' the code is not clean
- on '/@<name>' When we have 2 articles, if we click on any checkbox
it toggle the checkbox of the first article

Ydrasil committed:
Merge branch 'publication-status' of github.com:Ydrasil/neomad.org into publication-status

fuentesloic committed:
change simple by double quote

vinyll commented on blog/test.py:
```diff
@@ -165,3 +168,17 @@ def test_language_detection(self):
+                publication_date=None,
+                author=self.user).save()
+        result = self.client.get('/articles/')
+        self.assertNotIn(b'<article class=preview>', result.data)
```
I'd rather test that the title of the article does not appear, just like you do in line 177.

vinyll commented on blog/views.py:
```diff
@@ -11,7 +13,7 @@
 @app.route('/articles/')
 def article_list():
-    articles = Article.objects.all()
+    articles = Article.objects.filter(publication_date__ne=None)
```
You probably want to go for something like http://docs.mongoengine.org/guide/querying.html#default-document-queries.

vinyll commented on blog/views.py:
```diff
@@ -46,6 +48,10 @@ def article_create():
         else:
             article.title = request.form.get('title')
             article.content = clean_html(request.form.get('content'))
+            if request.form.get('publish') == 'on':
```
Relaying on the value `'on'` is pretty arbitrary as the value may change from the frontend when the frontend would evolve.
You should just: `if bool(request.form.get('publish')):…`
The responsability would be for the frontend to be coherent and send something that is [truthy of falsy](https://developer.mozilla.org/en/docs/Glossary/Truthy)

vinyll commented on core/views.py:
```diff
@@ -10,5 +10,5 @@
 @app.route('/')
 def home():
-    articles = Article.objects.all()[:3]
+    articles = Article.objects.all()[:3].filter(is_published__ne=False)
```
http://docs.mongoengine.org/guide/querying.html#default-document-queries would factorize your code as above

vinyll commented on static/css/_article_list.css:
```diff
@@ -56,3 +56,48 @@
+  top: 1.5rem;
+  width: 12rem;
+  background: #279E6A;
+  box-shadow: 0 0 0.3rem rgba(0,0,0,0.3);
```
spaces after commas

vinyll commented on static/css/edit.css:
```diff
@@ -152,3 +152,32 @@ main > header figure .loading.ready::before {
+  top: 1.5rem;
+  width: 16rem;
+  background: #279E6A;
+  box-shadow: 0 0 0.3rem rgba(0,0,0,0.3);
```
spaces after commas

vinyll commented on static/css/edit.css:
```diff
@@ -152,3 +152,32 @@ main > header figure .loading.ready::before {
+  left: 1.5rem;
+  top: 1.5rem;
+  width: 16rem;
+  background: #279E6A;
```
lower case or uppercase? Choice is yours, but stick with it ;)

vinyll commented on templates/blog/article.html:
```diff
@@ -39,6 +39,12 @@
     </ul>
   {% endif %}
 
+  {% if article.is_published==false %}
```
`{% if not article.is_published %}` is more read-friendly ;)

fuentesloic replied:
^^

vinyll commented on templates/blog/macros.html:
```diff
@@ -15,14 +15,19 @@
       {% else %}
       <time datetime={{ article.creation_date }} pubdate>{{ article.creation_date.strftime('%d %B, %Y') }}</time>
       {% endif %}
+      {% if not article.publication_date %}
```
{% if not article.is_published %}

vinyll commented on user/test.py:
```diff
@@ -48,8 +54,25 @@ def test_delete_trip(self):
+
+    def test_unpublished_articles_does_not_appears_if_the_author_not_logged(
+            self):
+        Article(title='<h1>Article that must not appear<br></h1>',
```
title with html??

vinyll commented on user/views.py:
```diff
@@ -14,10 +14,14 @@
+        if user == current_user:
+            articles = Article.objects(author=user)
+        else:
+            articles = Article.objects(author=user, publication_date__ne=None)
```
```python
articles = Article.objects(author=user)
if user != current_user:
    articles.filter(publication_date__ne=None)
```
Though refactoring with a queryset method would change this

fuentesloic committed:
fixed from vinyll review

vinyll commented:
Also bewarn of the commits. A manual interactive rebase might be appropriate here:
- for preserving who did what
- for renaming commits (some are still with a _WIP_ – work in progress – which obviously don't make sense for an old commit)
- for reducing 40something commits into a few
- for removing titles that don't make sense without their context

Ydrasil committed:
Add change from code review

Ydrasil committed:
Merge branch 'publication-status' of github.com:Ydrasil/neomad.org into publication-status

Ydrasil commented:
I have two idea for rebase interactively this branch, squash all the commit into one feature that is the publication status, or by keeping 1 commit with backend and 1 other for frontend.
What do you think about this solution and have you another one ?

vinyll commented:
If I were you, I would go for 1 or a couple of commit for each – the backend and the frontend – as you say, to preserve what each developer has done.

Ydrasil committed:
Merge branch 'master' into publication-status

fuentesloic committed:
Merge branch 'master' into publication-status

fuentesloic commented:
@Ydrasil could you squash commit plz ? (and I think you need a rebase) I would like merge it on master

Commit:
Merge branch 'master' into publication-status
